### PR TITLE
Chore: Stricter typing in type guards

### PR DESCRIPTION
**Pull request by joshhunt**
Removes `as` and `any` from type guards for stricter and safer code 😌.

Takes advantage of the recent typescript upgrade 😌

### CODE REVIEW

File: public/app/types/unified-alerting-dto.ts [hunk 1]
```diff
@@ -38,11 +38,19 @@ export function isGrafanaAlertState(state: string): state is GrafanaAlertState {
...
+
+  switch (state) {
+    case PromAlertingRuleState.Firing:
+    case PromAlertingRuleState.Inactive:
+    case PromAlertingRuleState.Pending:
+      return false;
+
+    default:
+      return true;
+  }
```
Comment (joshhunt) [inline public/app/types/unified-alerting-dto.ts:53]: not a fan of this change, will take another stab

File: packages/grafana-data/src/types/data.ts [hunk 2]
```diff
@@ -213,3 +213,27 @@ export interface DataConfigSource {
...
+  return value;
+}
+
+/**
+ * Utility type predicate to check if a value is typeof object, but excludes "null".
+ *
+ * We normally discourage the use of type predicates in favor of just inline typescript narrowing,
+ * but this is a special case to handle null annoyingly being typeof object
+ */
+export function isObject(value: unknown): value is object {
```
Comment (mckn) [inline packages/grafana-data/src/types/data.ts:231]: Any reason why we can't use `isObject` from lodash instead of implementing our own?

**Review by mckn — COMMENTED** [review]
LGTM! Suggesting that we are using `isObject` from lodash instead of implementing our own.

File: packages/grafana-data/src/types/data.ts (hunk 2 above)
Comment (joshhunt) [reply packages/grafana-data/src/types/data.ts:231]: it's different, this is one less dependency, and i trust this more - it's actually type checked!

File: public/app/features/alerting/unified/api/ruler.ts [hunk 3]
```diff
@@ -155,8 +156,13 @@ async function rulerGetRequest<T>(url: string, empty: T, params?: Record<string,
 }
 
 function isResponseError(error: unknown): error is FetchResponse<ErrorResponseMessage> {
-  const hasErrorMessage = (error as FetchResponse<ErrorResponseMessage>).data != null;
-  const hasErrorCode = Number.isFinite((error as FetchResponse<ErrorResponseMessage>).status);
+  if (!isObject(error)) {
+    return false;
+  }
+
+  const hasErrorMessage = 'data' in error && error.data !== null;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/api/ruler.ts:163]: Was the change to strict equality here intentional?

Comment (gillesdemey) [reply public/app/features/alerting/unified/api/ruler.ts:163]: Alrighty, can we change it to `error.data !== undefined && error.data !== null` then to keep the same behaviour as before?

File: packages/grafana-data/src/types/data.ts [hunk 4]
```diff
@@ -213,3 +213,27 @@ export interface DataConfigSource {
...
 type Truthy<T> = T extends false | '' | 0 | null | undefined ? never : T;
 export const isTruthy = <T>(value: T): value is Truthy<T> => Boolean(value);
+
+/**
+ * Serves no runtime purpose - only used to make typescript check a value has been correctly
+ * narrowed to an object
+ */
+function identityObject(value: object): object {
+  return value;
+}
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply packages/grafana-data/src/types/data.ts:223]: Is the jsdoc comment on both `identityObject` and `isObject` not clear enough?

User-defined type predicates are not type safe (it can return whatever it want, the narrowing is not type checked), and thus can easily be a source of introducing bugs and unpredictability.

`identityObject` is a function that typescript knows should only accept an argument of type `object`. Calling this in `isObject` is a build-time (not runtime) test that we have correctly narrowed `value` from `unknown` to `object`.

For example, if a developer were to come along and remove the `value !== null` check from `isObject`, they would immediately see a typescript error because now `object | null` would be passed into `identityObject`.

In other words, it serves no runtime purpose - only used to make typescript check a value has been correctly narrowed to an object.

[the remark this answers is not in the document]
Comment (mckn) [reply packages/grafana-data/src/types/data.ts:223]: Yeah, it makes sense. Could we fulfil the same thing with a couple of tests?

**Review by mckn — APPROVED** [review]
LGTM!
